src/graph/graphcontext.c

```diff
@@ -45,33 +37,76 @@ GraphContext *GraphContext_New(RedisModuleCtx *ctx, const char *graphname,
 
 	gc->string_mapping = array_new(char *, 64);
 	gc->attributes = raxNew();
-
+	Graph_SetMatrixPolicy(gc->g, SYNC_AND_MINIMIZE_SPACE);
 	QueryCtx_SetGraphCtx(gc);
 
-	// Set and close GraphContext key in Redis keyspace
-	RedisModule_ModuleTypeSetValue(key, GraphContextRedisModuleType, gc);
+	return gc;
+}
+
+/* This method tries to get a graph context, and if it does not exists, create a new one.
+ * The try-get-create flow is done when module global lock is acquired, to enforce consistency
+ * while BGSave is called. */
+GraphContext *_GraphContext_GetOrCreate(RedisModuleCtx *ctx, const char *graphname,
+										int readWriteFlag, size_t node_cap, size_t edge_cap) {
+	GraphContext *gc = NULL;
+	RedisModuleString *rs_name = RedisModule_CreateString(ctx, graphname, strlen(graphname));
+
+	// Acquire lock.
+	assert(pthread_mutex_lock(&_module_mutex) == 0);
+	// Open key with requested permission.
+	RedisModuleKey *key = RedisModule_OpenKey(ctx, rs_name, readWriteFlag);
+	if(RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+		// In case of non existing value, open key for write mode.
+		key = RedisModule_OpenKey(ctx, rs_name, REDISMODULE_WRITE);
+		// Create and initialize a graph context.
+		gc = _GraphContext_Create(graphname, node_cap, edge_cap);
+		// Set value in key.
+		RedisModule_ModuleTypeSetValue(key, GraphContextRedisModuleType, gc);
+		// Register graph context for BGSave.
+		GraphContext_RegisterWithModule(gc);
+		// Unlock.
+		assert(pthread_mutex_unlock(&_module_mutex) == 0);
+		goto cleanup;
+	} else {
+		assert(pthread_mutex_unlock(&_module_mutex) == 0);
+		if(RedisModule_ModuleTypeGetType(key) != GraphContextRedisModuleType) {
+			goto cleanup;
+		}
+		gc = RedisModule_ModuleTypeGetValue(key);
+		// Force GraphBLAS updates and resize matrices to node count by default
+		Graph_SetMatrixPolicy(gc->g, SYNC_AND_MINIMIZE_SPACE);
 
-	GraphContext_RegisterWithModule(gc);
+		QueryCtx_SetGraphCtx(gc);
+	}
 cleanup:
-	RedisModule_CloseKey(key);
 	RedisModule_FreeString(ctx, rs_name);
+	RedisModule_CloseKey(key);
 	return gc;
 }
 
+GraphContext *GraphContext_New(RedisModuleCtx *ctx, const char *graphname,
```

Comment:
seems a bit unnecessary, can't we have a single API, GraphContext_Retrieve which does get or create?

Reply:
Agree, this only has one invocation in cmd_bulk_insert.c now.